platform/linux-generic/pktio/dpdk.c

```diff
@@ -380,16 +374,30 @@ static struct rte_mempool *mbuf_pool_create(const char *name,
 	}
 
 	mbp_priv.mbuf_data_room_size = pool_entry->headroom +
-			pool_entry->seg_len;
+			pool_entry->seg_len + pool_entry->tailroom;
 	mbp_priv.mbuf_priv_size = RTE_ALIGN(sizeof(odp_packet_hdr_t),
 					    RTE_MBUF_PRIV_ALIGN);
 	rte_pktmbuf_pool_init(mp, &mbp_priv);
 
-	num = rte_mempool_populate_iova(mp, (char *)pool_entry->base_addr,
-					RTE_BAD_IOVA, pool_entry->shm_size,
-					NULL, NULL);
-	if (num <= 0) {
-		ODP_ERR("Failed to populate mempool: %d\n", num);
+	for (uint64_t i = 0; i < pool_entry->shm_size; i += page_size) {
+		remainder = (page_size - offset) % total_size;
+		addr = pool_entry->base_addr + i + offset;
+		ret = rte_mempool_populate_iova(mp, (char *)addr, rte_mem_virt2iova(addr),
+						page_size - remainder - offset,
+						NULL, NULL);
+
+		if (ret <= 0) {
+			ODP_ERR("Failed to populate mempool: %d\n", ret);
+			goto fail;
+		}
+
+		populated += ret;
+		offset = remainder ? total_size - remainder : 0;
+	}
+
+	if (populated != num) {
+		ODP_ERR("Failed to populate mempool with all requested blocks, populated: %u, "
+			"requested: %u\n", populated, num);
 		goto fail;
```

Comment:
As `rte_mempool_populate_iova()` itself doesn't take a count parameter, returns the amount of blocks it was able to map and there are several calls to it, it is safer to check that all buffers that will be used by ODP are actually also successfully mapped to DPDK side to avoid weird potential runtime issues.

Reply:
What I was after is that `rte_mempool_populate_iova()` documentation isn't that clear on the positive return value. So theoretically it could return >1 on success, which would cause this check to fail. However, this is unlikely and the issue would be easy to detect.